nixos/modules/services/networking/netmaker/netmaker.nix

```diff
@@ -0,0 +1,787 @@
+        serviceConfig.Type = "oneshot";
+        serviceConfig.RemainAfterExit = true;
+        script = lib.getExe (pkgs.writeShellApplication {
+          name = "netmaker-generate-config";
+          runtimeInputs = with pkgs; [
+            coreutils
+            gawk
+            gnugrep
+            gnused
+            systemd
+          ];
+          text = ''
+            MQ_PASSWORD_FILE="${cfg.mqtt.passwordFile}"
+
+            SECRETS="${envFile.generated.secrets}"
+            ENV="${envFile.generated.config}"
+
+            ${builtins.readFile ./netmaker-generate-config.sh}
+          '';
+        });
+      };
+
+      # based on https://github.com/gravitl/netclient/blob/b9ea9c9841f01297955b03c2b5bbf4b5139aa40c/daemon/systemd_linux.go#L14-L30
+      systemd.services.netmaker = {
+        description = "Netmaker Server daemon";
+        after = [ "network-online.target" "netmaker-configure.service" ];
+        wants = [ "network-online.target" ];
+        requires = [ "netmaker-configure.service" ];
+        wantedBy = [ "multi-user.target" ];
+
+        serviceConfig.EnvironmentFile = cfg.environmentFiles;
+        serviceConfig.WorkingDirectory = cfg.dataDir;
+        serviceConfig.ExecStart = "${lib.getExe cfg.package} -c ${cfg.config}";
+        serviceConfig.ExecStartPost = pkgs.writeShellScript "configure-nmctl" ''
+          export HOME=/root
+          ${lib.getExe cfg.nmctl.package} context set default \
+            --endpoint="https://${cfg.api.domain}" \
+            --master_key="$MASTER_KEY"
+        '';
+
+        serviceConfig.Restart = "on-failure";
+        serviceConfig.RestartSec = "15s";
+      };
+    }
+    {
+      # CoreDNS related config
+      services.coredns.enable = true;
+      # defines a snippet that can be integrated without clashes, see https://coredns.io/manual/toc/#reusable-snippets
+      services.coredns.config = ''
+        (netmaker) {
+          ${cfg.coredns.exposeAt}
+          hosts ${cfg.coredns.hostsPath} {
+            ttl 3600 # default
+            reload 5s
+            fallthrough
+          }
+        }
+        ${cfg.coredns.defaults.config}
+      '';
+      networking.firewall = with cfg.firewall.ports; {
+        allowedTCPPorts = [ cfg.coredns.internal.port ];
+        allowedUDPPorts = [ cfg.coredns.internal.port ];
+      };
+    }
+    (lib.mkIf (cfg.debugTools) {
+      environment.systemPackages = with pkgs; [
+        mqttui
+        (pkgs.writeShellApplication {
+          name = "netmaker-mqttui";
+          runtimeInputs = with pkgs; [ mqttui jq ];
+          text = ''
+            get() {
+              test -n "''${info:-}" || info="$(nmctl server info -o json | jq -c)"
+              jq -r "$@" <<<"$info"
+            }
+
+            export MQTTUI_PASSWORD="''${MQTTUI_PASSWORD:-"$(get '.MQPassword')"}"
+            export MQTTUI_USERNAME="''${MQTTUI_USERNAME:-"$(get '.MQUserName')"}"
+            export MQTTUI_BROKER="''${MQTTUI_BROKER:-"$(get '.Broker')"}"
```

Comment:
We don't have this information in nix code, do we?

Reply:
we have the username, but not the password which gets generated on the first run with configuration script so can as well pull the whole thing from the server.

Reply:
I considered sourcing the configuration files, but then it would include a lot more collateral envvars. I used it to debug actual `netclient` permissions so getting server config the same way as `netclient` made the most sense.